### ISSUE TITLE
Project direction

### ISSUE THREAD

**Issue by joonas-fi** [joonas-fi/promswarmconnect#2]
Let's discuss direction for this project, so that people possibly joining in know that we have common goals.

**Comment by joonas-fi** [comment]
Here are my desires:

- Feed Docker Swarm containers for the whole cluster to Prometheus
- Since this is somewhat critical part of the infrastructure, it should be stable enough (Glide or something for version pinning, testing where possible and set up a CI system for automatic builds). Keep code as simple and small as possible.
- Provide a Consul Catalog -compatible API so Prometheus (and other projects) can consume container data (service names, IPs, ports, labels etc.) out-of-the-box.
  * Labels from Docker to Consul would be required for automatic loadbalancing discovery for https://github.com/containous/traefik
- Keep it lean in RAM and CPU, so this can be ran on commodity DigitalOcean VMs and such

And nice to have features:

- Provide a way to scrape container & host metrics
  * maybe as a feature toggle
  * OR another project
  * OR refer them to Cadvisor if it is deemed better and doesn't consume too much resources
  * We should at least advice users on how to achieve this because most users probably want this

**Comment by byrnedo** [comment]
I'm all for the above.

For me the consul api is the big win because of `consul watch`.

With regards to dependencies, how do you feel about the following: 
- https://github.com/urfave/cli for cli (just makes life a whole lot easier with args. Not really required but nice)
- https://github.com/sirupsen/logrus for logging (sensible and widely used)
- https://github.com/cskr/pubsub for event handling: This one might need explaining:

I'm picturing one `eventwatcher` that distributes events pubsub style to our main handler, a `cataloger`, which holds our source of truth of endpoints/tasks, and then any plugins and then the web api. The need for it at the web api side would be to get the long polling events. There are other ways to do this but a pub sub is quite nice as long as we don't have a huge number of concurrent requests from the web api.

Edit:
Just looking though, does traefik support swarm mode natively? Seems to!

**Comment by byrnedo** [comment]
I threw together a draft of how the pub sub would work: https://github.com/byrnedo/prometheus-docker-swarm/tree/feature/pubsub
Makes things surprisingly simple, and of course brings some intellectual overhead too. 

Btw, don't take my enthusiasm for overbearing, I'm just doing this since I think it's fun, I'm not assuming that we agree on this

**Comment by joonas-fi** [comment]
I think I was getting ahead of myself, thinking that since we'd build a Docker-to-Consul bridge, we might as well support the use cases for Consul discovery like Traefik reading from Consul, but didn't think that yeah Traefik can tap off directly to Swarm as well.. :D This kind of thinking is dangerous, since the focus should first be in serving the core need: get metrics endpoints from Docker Swarm tasks to Prometheus. In the beginning we can forget container/host metrics as well... let's just get the core idea working flawlessly, and if that takes off then maybe think of the nice to haves. :)

If we both agree that we should integrate to Prometheus via the Consul API, then that's awesome and let's start exploring on that concept. Like you said, a good idea would be to first understand the protocol and what kind of HTTP requests Prometheus makes to Consul API. Setting up Prometheus, real Consul and getting a tcpdump of the API interaction would be a great start in reverse engineering it (I found it a bit difficult to understand the protocol just by looking at Consul's source code - but perhaps I'm a bit simple :) ).

I'm using Glide in my other projects as well - it seems to be the de facto standard until `go dep` is officially released.

Logrus seems to be widely used so it's awesome.

https://github.com/urfave/cli seems to have more stars than https://github.com/spf13/cobra, but Cobra seems to be embraced by highly respected projects so if we need a CLI library, I'd vote for that.

I have no objections for a pub/sub system if we really need it (I understood that it's for a multi-recipient delivery in contrast to a "random" recipient with bare Golang channels if there are multiple consumers). I'm having a hard time understanding the need for multiple listeners for a topic, but perhaps I'll get it when I'll look at your `pubsub` branch soon. :)

**Comment by byrnedo** [comment]
Agreed. Basics first :)

I've seen cobra a lot but never used it, I can go for that.

Yeah, the multiple recipient possibilities are the reason. I was initially thinking of the consul long polling there, but it actually might tend itself well to writing plugins/extension since you'd just need to subscribe to a channel for events like 'endpoint.created' or 'endpoint.remove'.

**Comment by byrnedo** [comment]
I think consul could be stage 2 really. A more robust service discovery for prometheus is definitely goal #1 for now.

**Comment by byrnedo** [comment]
Did you ever find a way to allow the metrics path to be configurable btw? I see that it's set in the conf itself which I'm guessing one can't just generate with good conscience.

**Comment by joonas-fi** [comment]
> Did you ever find a way to allow the metrics path to be configurable btw? I see that it's set in the conf itself which I'm guessing one can't just generate with good conscience.

That's a good question!

The `metrics_path` in the Prometheus configuration is just a default, unless it's overridden by a target. If you provide a `__metrics_path__` label to Prometheus then that is used (https://prometheus.io/docs/operating/configuration/#relabel_config). I think I just didn't have time to implement the regex parsing for the path part in https://github.com/function61/prometheus-docker-swarm/blob/master/utils.go#L62 because I didn't know how to use capture groups with Golang regex yet :D (I think this literally was the first project I started to write in Golang).

The resulting file should look like this:

```
[
	{
		"targets": [
			"10.0.0.23:80"
		],
		"labels": {
			"job": "html2pdf",
			"__metrics_path__": "/foometrics"
		}
	}
]
```

I think we should populate that `__metrics_path__` for clarity regardless of whether the path is default (`/metrics`) or not, so the file structure is always uniform and there's a clear mapping from Swarm's `METRICS_ENDPOINT` ENV var.

**Comment by joonas-fi** [comment]
> I think consul could be stage 2 really. A more robust service discovery for prometheus is definitely goal #1 for now.

I'd like to hear your thoughs on this. In my mind the robustness is achieved by emulating the Consul API.

If we don't use it, we have to to integrate via the file config option. That means that we must be able to write to the filesystem on the Prometheus process, either by:

- Running this "bridge" binary in the same container
  * This is what we're doing now, and is dangerous because if you look at https://github.com/function61/prometheus-docker-swarm/blob/master/run.sh#L5, you'll see that if the bridge process dies, nobody will restart it.
  * I wouldn't want to have to include a process manager to guarantee that they're both running. IIRC Prometheus is a statically compiled binary, and because of that *the process manager binary* has to be compiled statically (we do it here: https://github.com/function61/prometheus-docker-swarm/blob/master/Makefile#L6 ) to work in Prometheus' container because there is no libc available (the only dependency of Golang-based binaries), and the only lightweight procman I found was https://github.com/mattn/goreman which requires libc unless we compile it ourselves..
  * And besides running multiple binaries in a single container is not the "Docker way" IMO.
  * Also this hack is not good because the bridge and Prometheus versions cannot be upgraded separately, but instead for every Prometheus version we'd have to publish this bridge image because they're so deeply intertwined (our bridge image contains the Prometheus image).
- Or run in a separate container but still for the file writing part we'd have to have a bind mount that is seen by both containers (Prometheus and the bridge), which constrains both containers to reside in the same host machine, which requires you to specify placement constraints for these two Swarm Tasks... I don't like this.

So my thought process was: the bridge needs to be run in a separate container, and possibly on a different node => the integration needs to be over HTTP => the best out-of-the-box integration that works over HTTP was Consul.

WDYT?

edit: simply put: even if you write the most perfect code in the world to make the bridge binary robust, it will crash someday because the server runs out of RAM or some other crap like that. So the process will die, so somebody will have to restart it. We could hack our way by using a process manager, but Docker itself is a process manager so we should delegate process restarts to the infrastructure - the same infrastructure that is designed keep your other containers running (taking process crashes/restarts to our responsibility would be re-inventing a wheel and we'd probably do a bad job of it).

**Comment by byrnedo** [comment]
Awesome with the explanation about the labels, had no idea they could be overwritten in the targets! :)

I do agree that without some http api then it's really clumsy. And I suppose the version you have right now does the file job simply and well even if it's polling. At least it's correct all the time, and getting millisecond precision on target metrics doesn't in my mind matter.

In my mind I would probably always run the file version with a mount shared by prometheus but it's hardly elegant or future-proof.

I also think that consul is probably a good a way as any to go. As we've said before it will make this more than just a docker prom bridge.

I guess my reason on it being a stage 2 was more just commitment and complexity, but in hindsight I think it's worth it really (and probably quite interesting).

> If we both agree that we should integrate to Prometheus via the Consul API, then that's awesome and let's start exploring on that concept. Like you said, a good idea would be to first understand the protocol and what kind of HTTP requests Prometheus makes to Consul API. Setting up Prometheus, real Consul and getting a tcpdump of the API interaction would be a great start in reverse engineering it (I found it a bit difficult to understand the protocol just by looking at Consul's source code - but perhaps I'm a bit simple :) ).

At this point I definitely agree! I also tried looking through the code but it was not easy so I think you've the best idea. Also, one sneaky call to the kv store might change everything so best to get a detail view of the requests.

Shall we make a card with some investigation tasks for this, I kind of R&D to see if it's worth the trouble or possible?

**Comment by byrnedo** [comment]
Also, I just saw that actually the events that are swarm global are only for service create, update, remove. There still is a better way than polling the api, if on service change we start polling the internal dns on `tasks.<service-net-alias>`. Would be a lot quicker and less demanding on the docker engine.

Edit: Another option might be to figure out the task ips based on the netmask and replicas (assuming they're predictably allocated)

**Comment by joonas-fi** [comment]
Sorry for the slow reply, have been a bit busy lately!

I took a look at your "pubsub" branch. You seem to have removed the pub/sub feature now :D But I see the design it used to have on the older code tree when it still used to exist. It looked actually pretty nice. :) But of course if pub/sub is not required (I guess since you removed it) then less is simpler. :)

I also noticed that you started putting code in different packages (such as the plugins), which I like because I've been a fan of splitting separate concepts into their own packages where it makes sense to.. I'm not sure what the Golang ecosystems' vision of this is because I've seen many apps where packages have large amounts of files with obvious different concepts living in separate files but same package (= they can in theory depend on each other which I don't like), but to me many packages feels "the right way". :)

> In my mind I would probably always run the file version with a mount shared by prometheus but it's hardly elegant or future-proof.

Yeah I don't like it because it forces us to pin Prometheus and this bridge to run on the same, one explicit node. And that makes us more prone to failure because if the node running Prometheus dies, Swarm cannot just re-schedule Prometheus (or the bridge) to run on a different node. Ok in practice you still need to pin Prometheus to a node because it has state (unless you use Flocker or something for handling/migrating state) which we don't want to vanish if the container is restarted, but I wouldn't want to have one more reason to make container placement less flexible.

Also, before we waste more of your time I hope you're aware that if our goal is just to enable service discovery from Swarm in Prometheus, you could also use your effort and do it directly in Prometheus as a SD plugin. And the whole "emulate Consul" approach is just to support plugging Swarm data in without modifying Prometheus (which is not something we'd have to do if Swarm SD was written directly inside Prometheus). My goal was to make a proof-of-concept outside of Prometheus because it was quicker to code than learning how to develop against Prometheus' codebase, and maybe I was dreaming of integrating host/container metrics gathering in this one as well to have fewer and simpler/leaner components to operate than to add cadvisor and stuff. :)

That being said, if you still think that your time is well spent in this design (Consul-Swarm bridge outside of Prometheus with the added promise of more stuff like host/container metrics), I'm glad to work with you and give write access to the repo. :)

> Also, I just saw that actually the events that are swarm global are only for service create, update, remove. 

Darn it... Oh well back to polling then. I suppose via DNS is more efficient than polling for Docker's REST API like you said. And thanks for doing good research!

> Another option might be to figure out the task ips based on the netmask and replicas (assuming they're predictably allocated)

I don't know how documented that would be, but in general it's not a good idea to rely on implementation details, unless they're clearly documented and promised that they're not going to change, and we don't have to code for all the surprising corner cases.. :)

I can't promise anything because this week looks a bit busy as well, but I've been thinking a bit about the setup of reverse engineering Consul's API: spin a fresh Consul, fresh Prometheus with it configured to connect to Consul's API. Register a task in Consul (and verify it shows up in Prometheus), then take a tcpdump capture and in Wireshark see the API requests it makes to Consul.. then document it here so we can learn together. I just don't have the time to set that up right now, but I could probably soon do it.

Also, I noticed that there's a [Marathon SD](https://prometheus.io/docs/operating/configuration/#marathon_sd_config) plugin whose API URL is configurable (some SD plugins are for SaaS platforms) and looks a lot simpler (uses polling instead of long polling) which is an API we probably could mock more easily, but then again it's not Consul (and therefore it's less usable if we'd be thinking about more projects benefiting from a generic Swarm-to-Consul bridge). But we could consider Marathon emulation as a quick proof-of-concept if it would benefit us before committing to the Consul part..

**Comment by byrnedo** [comment]
No worries at all, I'm the same!

About the pubsub: I got worried just about trying to avoid deadlocks and didn't really know the full capabilities of that little library so thought that simpler was better for now.

And with the packages; yeah, I'm not quite sure what best practice is with that. Maybe having larger packages promotes writing more encapsulated code (structs with methods vs plain functions).

> Also, before we waste more of your time I hope you're aware that if our goal is just to enable service discovery from Swarm in Prometheus, you could also use your effort and do it directly in Prometheus as a SD plugin.

Thats a good point. I guess personally I was most interested in getting somthing that would provide targets to prometheus but also for templating other services. In the case of an load balancer it would be much better to be more frequent than polling I would think. But after you pointed me towards traefik, and after having a look at openresty, I think maybe I'd better just go with one of those.

Since docker doesn't support swarm task events (yet - there is some vague talk that it might be implemented, they are reluctant due to the amount of data that would be sent), the best one could do would be watching the dns entries. Downside is one wont get access to unhealthy targets AFAIK. Which feels a bit limiting.

My feeling now is to wait until or if docker starts shipping with task events, then make the POC into a reliable and fast termplater. Then after that is done, and we have a solid base, look at implementing some prometheus compatible api.

**Comment by joonas-fi** [comment]
@byrnedo just a heads up, I made some progress: https://github.com/prometheus/prometheus/issues/1766#issuecomment-448962898

I realize that the mind share has probably moved to Kubernetes, but I haven't had the time to learn it and migrate to there. I'm now running promswarmconnect in production and it seems to work quite well.

**Comment by joonas-fi** [comment]
@byrnedo I don't know where your response got lost (I got it in email), but I'm not seeing it anywhere. I pasted it here:

> nice :) I've not been doing much on swarm lately but after having used k8s I might start doing so again. K8s is reallllllly heavy. I like how light swarm is.

I'm curious to hear about your experiences, since I'm interested about learning about k8s! In what regard do you consider it "heavy" - is it mentally hard to understand, or is it heavy on the server resources like RAM and CPU? Both?

I'm currently running a super small startup, so having the ability to use cheaper VMs with low overhead of the infrastructure layer is important to me.

**Comment by byrnedo** [comment]
Hey @joonas-fi , reason it disappeared was that I replied on the wrong repo :p Then forgot to post it here.

Mentally I think you're pretty much prepared if you've run a production setup on swarm. I had put together things that come for free with k8s (highly configurable and reactive ingress controller/reverse proxy). 
I found out pretty quickly though that the version that kops (if you haven't already, check it out, it's googles provisioning cli for k8s) provisions needs like 3/4 fairly big amazon instances ( I forget which exactly), and that idle it was costing 400usd a month.

I don't mean that kubernetes is in any way bad. It offers waaaay more than swarm. It's just that I don't think I would use it for every project. I would use it where there's a bigger team involved, bigger budget too. 

If I was back at a startup I'd probably start with swarm and move to k8s later on.

**Comment by byrnedo** [comment]
Also, it would depend on my colleagues. If everyone was interested on getting their hands dirty, swarm could work. But if people just want it to work and do everything that's needed then k8s would be better.

**Comment by joonas-fi** [comment]
Ahh haha that explains.. I replied to the email notification, so by extension I also replied to the wrong repo 🤦‍♂️  😆  I was wondering why that happened, but now I know. :)

Oh, it's cool that k18s has reverse proxying, because with Swarm you pretty much have to bolt that on.. the port-based loadbalancing in Swarm doesn't get you far unless you're only running one HTTP service in your cluster! I'm using Traefik and i has worked great for me. Zero configuration which is great for a lazy bastard like me!

The default instances kops provisioned for you sound pretty beefy and expensive machines.. I'm left wondering what is the minimal recommended setup for k18s and the overhead of k18s regarding CPU/RAM.. I've tried Googling but haven't found anything clear. Perhaps it's not a problem since people are not talking about the overhead with cheap VMs..

p.s. I am closing this issue, since I think my design for this is "final" :)
